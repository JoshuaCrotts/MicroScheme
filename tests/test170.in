;
; Account OOP design
;
(define cadr (lambda (l) (car (cdr l))))
(define (make-account)
   (let ((my-balance 0))
      (lambda (args)
          (cond ((eq? (car args) 'balance) my-balance)
                ((eq? (car args) 'withdraw)
                    (if (>= my-balance (cadr args))
                       (begin (set! my-balance (- my-balance (cadr args)))
                               my-balance)
                       "Insufficient funds"))
                ((eq? (car args) 'deposit)
                    (begin
                        (set! my-balance (+ my-balance (cadr args)))
                        my-balance))
                (else "unknown request")))))

(define account-1 (make-account))
(account-1 '(deposit 300))
(account-1 '(balance))
(account-1 '(withdraw 175))
(account-1 '(balance))
(define account-2 (make-account))
(account-2 '(deposit 1000))
(account-2 '(balance))
(account-2 '(deposit 215))
(account-2 '(balance))
(account-2 '(withdraw 1065))
(account-1 '(balance))
(account-2 '(balance))