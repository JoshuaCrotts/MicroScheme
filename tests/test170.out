300
300
125
125
1000
1000
1215
1215
150
125
150
